live/liveview.go

```diff
@@ -23,22 +23,124 @@ import (
 
 // Config is the configuration for a live application.
 type Config struct {
-	// LayoutTemplate is a template that wraps all Views
-	LayoutTemplate *htmltmpl.Template
-	// Mux is a http.Handler that routes requests to Views
+	// Mux is a http.Handler that routes requests to Views.
 	Mux http.Handler
-	// PageTitleConfig is a configuration for the page title if the application
-	// uses the liveTitleTag template function in its LayoutTemplate.
-	PageTitleConfig PageTitleConfig
+	// WriteLayout is a func that writes your base layout HTML to a response writer for a given request and dot.
+	// You are responsible for carrying the contents of the LiveViewDot through to your layout template.
+	// If your layout template uses the {{ liveViewContainerTag }} func, it should be passed the contents of LiveViewDot as its argument.
+	WriteLayout func(http.ResponseWriter, *http.Request, *LiveViewDot) error
 }
 
-func (c *Config) viewForRequest(r *http.Request, currentView View) (View, error) {
-	rw := &JoinHandler{
+type liveViewRequestContextKey struct{}
+type liveViewContainer struct {
+	lv View
+	r  *http.Request
+}
+
+func (c *Config) viewForRequest(w http.ResponseWriter, r *http.Request, currentView View) (View, int, *http.Request) {
+	container := &liveViewContainer{
 		lv: currentView,
 	}
+	r = r.WithContext(context.WithValue(r.Context(), liveViewRequestContextKey{}, container))
+
+	rw := &joinHandler{
+		w: w,
+	}
 	c.Mux.ServeHTTP(rw, r)
-	// TODO: check rw.code, convert some status codes (500s?) into meaningful errors
-	return rw.lv, nil
+	if container.r != nil {
+		r = container.r
+	}
+	return container.lv, rw.code, r
+}
+
+func (c *Config) Middleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// Check for cases where our config's router is its own Mux.
+		// In that case we will re-enter the middleware and should no-op.
+		_, ok := w.(*joinHandler)
+		if ok {
+			next.ServeHTTP(w, r)
+			return
+		}
+		// Get the LiveView if one is routable.
+		lv, code, r := c.viewForRequest(w, r, nil)
+
+		// If the inner router 500s, cease the middleware chain.
+		if code%100 == 5 {
+			return
+		}
+
+		// If no view was found continue the chain without upgrading the request to a live one;
+		// the outer router will presumably serve this route, but we no longer care about it.
+		if lv == nil {
+			next.ServeHTTP(w, r)
+			return
+		}
+
+		// At this point we know this is a "live" route.
+		// Configure things with our view and call the appropriate lifecycle methods.
+
+		// if View implements HasPageTitleConfig interface then
```

Comment:
Do we still need this, now that we use a callback function?  They can set the page title themselves however they please, without having to threaded back and forth through LiveViewDot.

Reply:
I think we still need this for live-updating the title via websocket updates.

Reply:
Wait, I was wrong—this isn't part of websocket-land. I'll remove it.

Reply:
Nope, I was right the first time: we even support doing this already, I just didn't realize it.